File: frontend/src/features/projects/types/trace.ts
export interface ProjectTraceEvent {
  readonly id: string;
  readonly projectId: string;
  readonly threadId: string | null;
  readonly eventType: string;
  readonly payload: Record<string, unknown>;
  readonly createdAt: string;
}

export interface ProjectTraceEventsResponse {
  readonly events: readonly ProjectTraceEvent[];
}

File: backend/app/features/projects/contracts/trace.py
"""Contracts for project trace timelines."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class _TraceModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class ProjectTraceEventContract(_TraceModel):
    id: str
    project_id: str = Field(alias="projectId")
    thread_id: str | None = Field(default=None, alias="threadId")
    event_type: str = Field(alias="eventType")
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: str = Field(alias="createdAt")


class ProjectTraceEventsResponse(_TraceModel):
    events: list[ProjectTraceEventContract] = Field(default_factory=list)
